python/seldon_core/microservice.py

```diff
@@ -351,107 +333,108 @@ def main():
-            server1_func = rest_prediction_server
-
-    elif args.api_type == "GRPC":
-        seldon_metrics = SeldonMetrics(
-            worker_id_func=lambda: threading.current_thread().name
+        logger.info(
+            "REST microservice running on port %i single-threaded=%s",
+            http_port,
+            args.single_threaded,
         )
+        server1_func = rest_prediction_server
+    else:
+        # Start production server
+        def rest_prediction_server():
+            options = {
+                "bind": "%s:%s" % ("0.0.0.0", http_port),
+                "accesslog": accesslog(args.log_level),
+                "loglevel": args.log_level.lower(),
+                "timeout": 5000,
+                "threads": threads(args.threads, args.single_threaded),
+                "workers": args.workers,
+                "max_requests": args.max_requests,
+                "max_requests_jitter": args.max_requests_jitter,
+                "post_worker_init": post_worker_init,
+            }
+            if args.pidfile is not None:
+                options["pidfile"] = args.pidfile
+            app = seldon_microservice.get_rest_microservice(user_object, seldon_metrics)
 
-        def grpc_prediction_server():
+            UserModelApplication(
+                app,
+                user_object,
+                jaeger_extra_tags,
+                args.interface_name,
+                options=options,
+            ).run()
 
-            if args.tracing:
-                from grpc_opentracing import open_tracing_server_interceptor
+        logger.info("REST gunicorn microservice running on port %i", http_port)
+        server1_func = rest_prediction_server
 
-                logger.info("Adding tracer")
-                interceptor = open_tracing_server_interceptor(tracer)
-            else:
-                interceptor = None
+    def grpc_prediction_server():
 
-            server = seldon_microservice.get_grpc_server(
-                user_object,
-                seldon_metrics,
-                annotations=annotations,
-                trace_interceptor=interceptor,
-            )
+        if args.tracing:
+            from grpc_opentracing import open_tracing_server_interceptor
 
-            try:
-                user_object.load()
-            except (NotImplementedError, AttributeError):
-                pass
+            logger.info("Adding tracer")
+            tracer = setup_tracing(args.interface_name)
+            interceptor = open_tracing_server_interceptor(tracer)
+        else:
+            interceptor = None
 
-            server.add_insecure_port(f"0.0.0.0:{port}")
+        server = seldon_microservice.get_grpc_server(
+            user_object,
+            seldon_metrics,
+            annotations=annotations,
+            trace_interceptor=interceptor,
+        )
 
-            server.start()
+        try:
+            user_object.load()
```

Comment:
Would the load function not be called quite a few times across each of the respective functions repeatedly? Perhaps it could just be called if the ready variable not set

Reply:
Suggest a separate issue for this?